### ISSUE TITLE
go-to-definition does not always work correctly on using statements

### ISSUE THREAD

**Issue by kirillbobyrev** [clangd/clangd#480]
Consider the following example:

```c++
namespace ns {
void foo(int) {}
class Foo {};
}

using ns::foo;
//        [0]
using ns::Foo;
//        [1]

int main() {
  foo(5);
//[2][3]
  Foo f;
//[4]
}
```

There are 5 ranges (`0` through `4`, each starts with `[` and ends with `]` including the bracket locations). Calling "go to definition" on each of these ranges has the following outcome:

0. Goes to the same `using ns::foo;` statement it's called on, pointing to the beginning of range 0.
1. Goes to the same `using ns::Foo;` statement it's called on, pointing to the beginning of range 1.
2. Goes to `using ns::foo;`, points to the beginning of range 0.
3. Goes to `void foo(T t) {}`.
4. Goes to `class Foo {};`.

I think the problems with this behaviour are:

1. Calling "go to definition" on `using ns::foo` and `using ns::Foo` (ranges 0 and 1) should go to the function and class definitions, not to the using statements (it's not really useful otherwise).
2. Calling "go to definition" on foo (range 2) should go to the function definition, not the using statement (range 0). This is arguable though.
3. "go to definition" on both function name and brackets (ranges 2 and 3) + argument list should be consistent with each other: it's quite confusing that they go to different locations now.
4. "go to definition" on ranges 3 and 4 (function name resolved through `using` and class name resolved through `using`) should be consistent; it's not clear why behaviour is different for functions and classes.

**Comment by hokein** [comment]
Yeah, this is similar to #277, this reminds me of a proposed fix (https://reviews.llvm.org/D74054), I somehow forgot it.

**Comment by kirillbobyrev** [comment]
Ah, you're right! I think this covers case 1, right? Would 2, 3, 4 still be valid additions and make sense, too?

**Comment by kirillbobyrev** [comment]
FWIW I think the "references" request on type alias works fine.

**Comment by hokein** [comment]
> I think this covers case 1, right? Would 2, 3, 4 still be valid additions and make sense, too?

yeah, 1 is the most interesting case. 2, 3, 4 seems reasonable to me, I think go-to-def's behavior should be at least consistent on functions and classes. The proposed patch fixes 1 and also makes the go-to-def behavior consistent.
